cpp/ql/test/experimental/query-tests/Security/CWE/CWE-476/semmle/tests/test.cpp

```diff
@@ -0,0 +1,252 @@
+void* operator new[](size_t, std::align_val_t, float);
+void operator delete(void*, float);
+void operator delete[](void*, float);
+void operator delete(void*, std::align_val_t, float);
+void operator delete[](void*, std::align_val_t, float);
+
+struct myData
+{
+    int		sizeInt;
+    char*	buffer;
+};
+
+struct myGlobalData
+{
+    int		sizeInt;
+    myData**	bufMyData;
+};
+
+void allocData(myGlobalData * dataP) {
+    for (size_t i = 0; i < dataP->sizeInt; i++)
+    {
+        dataP->bufMyData[i] = new myData;
+	dataP->bufMyData[i]->sizeInt = 10;
+	dataP->bufMyData[i]->buffer = new char[dataP->bufMyData[i]->sizeInt];
+    }
+}
+
+
+void throwFunction(int a) {
+	if (a == 5)  throw "my exception!";
+}
+void throwFunction2(int a) {
+	if (a == 5) throw exception();
+}
+void funcWork1b() {
+  int a;
+  myGlobalData *valData = new myGlobalData;
+
+  try { 
+  	cleanFunction();
+	throwFunction(a);
+	valData->sizeInt  = 10;
+        valData->bufMyData =  new myData*[valData->sizeInt];
+	cleanFunction();
+	allocData(valData);
+	cleanFunction();
+	 
+  } 
+  catch (...) 
+  {
+    for (size_t i = 0; i < valData->sizeInt; i++)
+    {
+    	delete[] valData->bufMyData[i]->buffer; // BAD
+        delete valData->bufMyData[i];
+    }
+    delete [] valData->bufMyData;
+    delete valData;
+  }
+}
+
+void funcWork1() {
+  int a;
+  myGlobalData *valData = new myGlobalData;
+  valData->sizeInt  = 10;
+  valData->bufMyData =  new myData*[valData->sizeInt];
+  try { 
+  	cleanFunction();
+	throwFunction(a);
+	cleanFunction();
+	allocData(valData);
+	cleanFunction();
+	 
+  } 
+  catch (...) 
+  {
+    for (size_t i = 0; i < valData->sizeInt; i++)
+    {
+	if (valData->bufMyData[i]) 
+       	    delete[] valData->bufMyData[i]->buffer; // GOOD
+        delete valData->bufMyData[i];
+    }
+    delete [] valData->bufMyData;
+    delete valData;
+  }
+}
+
+void funcWork2() {
+  int a;
+  myGlobalData *valData = new myGlobalData;
+  valData->sizeInt  = 10;
+  valData->bufMyData =  new myData*[valData->sizeInt];
+  try { 
+ 	do {
+		cleanFunction();
+		allocData(valData);
+		cleanFunction();
+		throwFunction(a);
+
+	}
+	while(0);
+	 
+  } 
+  catch (...) 
+  {
+    for (size_t i = 0; i < valData->sizeInt; i++)
+    {
+    	delete[] valData->bufMyData[i]->buffer; // GOOD
+        delete valData->bufMyData[i];
+    }
+    delete [] valData->bufMyData;
+    delete valData;
+  }
+}
+void funcWork3() {
+  int a;
+  myGlobalData *valData = new myGlobalData;
+  valData->sizeInt  = 10;
+  valData->bufMyData =  new myData*[valData->sizeInt];
+  try { 
+	cleanFunction();
+	allocData(valData);
+	cleanFunction();
+	throwFunction(a);
+ 
+  } 
+  catch (...) 
+  {
+    for (size_t i = 0; i < valData->sizeInt; i++)
+    {
+    	delete[] valData->bufMyData[i]->buffer; // GOOD
+        delete valData->bufMyData[i];
+    }
+    delete [] valData->bufMyData;
+    delete valData;
+  }
+}
+
+
+void funcWork4() {
+  int a;
+  myGlobalData *valData;
```

Comment:
```suggestion
  myGlobalData *valData = 0;
```
because the `new` on line 156 could fail, taking us straight to the `catch` block.